File: src/llm/grouping.js
import { sha256Of } from '../util/hash.js';

/**
 * Request grouping and partitioning (design section 5.1; flow step 6.2).
 *
 * Alerts are grouped by `alert_rule_url` where present and by `application` where it is
 * not. Each group is sent independently: groups are NEVER packed together to fill
 * capacity, even when they are small, so a request never mixes alert rules or
 * applications.
 *
 * The bet this rests on: showing a rule's instances together is context a reviewer would
 * want - the model can see whether a message is genuinely per-instance or one generic
 * string repeated across forty nodes. Design section 7.1 records that this is still
 * unvalidated, which is why grouping is isolated here and testable on its own.
 */

/** Hard ceiling from design section 5.1. Configuration may lower it, never raise it. */
export const MAX_BATCH_SIZE = 200;

/**
 * @typedef {import('../domain/normalize.js').AlertRecord} AlertRecord
 */

/**
 * @typedef {object} AlertGroup
 * @property {'alert_rule_url'|'application'} type
 * @property {string} value
 * @property {AlertRecord[]} alerts Sorted deterministically.
 */

/**
 * @typedef {object} Batch
 * @property {string} batchId
 * @property {'alert_rule_url'|'application'} groupType
 * @property {string} groupValue
 * @property {number} partitionIndex
 * @property {number} partitionCount
 * @property {AlertRecord[]} alerts
 * @property {string[]} alertIds Parallel to `alerts`.
 */

/**
 * Transport identifier for one alert.
 *
 * Explicitly NOT a second business identity: it is the SHA-256 of an unambiguous encoding
 * of (schema, application, key_field). The verdict cache stays keyed on
 * (application, key_field, prompt_version, model_version) as decided.
 *
 * @param {AlertRecord} alert
 * @returns {string}
 */
export function alertTransportId(alert) {
  return sha256Of({
    schema: alert.schema,
    application: alert.application,
    key_field: alert.keyField,
  });
}

/**
 * Deterministic batch identifier.
 *
 * @param {object} parts
 * @param {string} parts.runId
 * @param {string} parts.groupType
 * @param {string} parts.groupValue
 * @param {number} parts.partitionIndex
 * @param {string[]} parts.alertIds Ordered exactly as sent.
 * @param {string} parts.promptVersion
 * @param {string} parts.modelVersion
 * @returns {string}
 */
export function batchIdOf(parts) {
  return sha256Of({
    run_id: parts.runId,
    group_type: parts.groupType,
    group_value: parts.groupValue,
    partition_index: parts.partitionIndex,
    alert_ids: parts.alertIds,
    prompt_version: parts.promptVersion,
    model_version: parts.modelVersion,
  });
}

/**
 * Is a value usable as a grouping key?
 * @param {unknown} value
 * @returns {boolean}
 */
function isUsable(value) {
  return typeof value === 'string' && value.trim() !== '';
}

/**
 * Group alerts by rule URL, falling back to application.
 *
 * API alerts do not carry an alert-rule URL at all, which is why the fallback exists and
 * why R4 does not penalise them for the absence.
 *
 * @param {AlertRecord[]} alerts
 * @returns {AlertGroup[]} Ordered by (type, value).
 */
export function groupAlerts(alerts) {
  /** @type {Map<string, AlertGroup>} */
  const groups = new Map();

  for (const alert of alerts) {
    const type = isUsable(alert.alertRuleUrl) ? 'alert_rule_url' : 'application';
    const value =
      type === 'alert_rule_url' ? /** @type {string} */ (alert.alertRuleUrl) : alert.application;
    const key = `${type}${value}`;
    let group = groups.get(key);
    if (!group) {
      group = { type, value, alerts: [] };
      groups.set(key, group);
    }
    group.alerts.push(alert);
  }

  for (const group of groups.values()) {
    // Within a group, alerts are ordered by key_field; application breaks the tie so the
    // order is total even when two applications share a key_field value.
    group.alerts.sort(
      (a, b) =>
        compareStrings(a.keyField, b.keyField) || compareStrings(a.application, b.application),
    );
  }

  return [...groups.values()].sort(
    (a, b) => compareStrings(a.type, b.type) || compareStrings(a.value, b.value),
  );
}

/**
 * @param {string} a
 * @param {string} b
 * @returns {number}
 */
function compareStrings(a, b) {
  return a < b ? -1 : a > b ? 1 : 0;
}

/**
 * Split `n` items into `ceil(n / max)` balanced partitions whose sizes differ by at most
 * one.
 *
 * 401 alerts become 134, 134 and 133 - never 200, 200 and 1. A tiny tail request would
 * lose exactly the same-group context that batching exists to provide.
 *
 * @param {number} n
 * @param {number} max
 * @returns {number[]} Partition sizes, largest first.
 */
export function balancedPartitionSizes(n, max) {
  if (n <= 0) return [];
  const count = Math.ceil(n / max);
  const base = Math.floor(n / count);
  const remainder = n % count;
  return Array.from({ length: count }, (_, i) => (i < remainder ? base + 1 : base));
}

/**
 * Build the batches for one run.
 *
 * @param {AlertRecord[]} alerts Alerts eligible for assessment.
 * @param {object} options
 * @param {string} options.runId
 * @param {string} options.promptVersion
 * @param {string} options.modelVersion
 * @param {number} [options.maxBatchSize]
 * @returns {Batch[]}
 */
export function buildBatches(alerts, options) {
  const maxBatchSize = Math.min(options.maxBatchSize ?? MAX_BATCH_SIZE, MAX_BATCH_SIZE);
  /** @type {Batch[]} */
  const batches = [];

  for (const group of groupAlerts(alerts)) {
    const sizes = balancedPartitionSizes(group.alerts.length, maxBatchSize);
    let offset = 0;
    sizes.forEach((size, partitionIndex) => {
      const slice = group.alerts.slice(offset, offset + size);
      offset += size;
      const alertIds = slice.map((a) => alertTransportId(a));
      batches.push({
        batchId: batchIdOf({
          runId: options.runId,
          groupType: group.type,
          groupValue: group.value,
          partitionIndex,
          alertIds,
          promptVersion: options.promptVersion,
          modelVersion: options.modelVersion,
        }),
        groupType: group.type,
        groupValue: group.value,
        partitionIndex,
        partitionCount: sizes.length,
        alerts: slice,
        alertIds,
      });
    });
  }

  return batches;
}

File: src/llm/grouping.py
"""Request grouping and partitioning (design section 5.1; flow step 6.2).

Alerts are grouped by ``alert_rule_url`` where present and by ``application`` where it is
not. Each group is sent independently: groups are NEVER packed together to fill capacity,
even when they are small, so a request never mixes alert rules or applications.

The bet this rests on: showing a rule's instances together is context a reviewer would want
- the model can see whether a message is genuinely per-instance or one generic string
repeated across forty nodes. Design section 7.1 records that this is still unvalidated,
which is why grouping is isolated here and testable on its own.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Final

from src.domain.normalize import AlertRecord
from src.hashing import sha256_of

__all__ = [
    "MAX_BATCH_SIZE",
    "AlertGroup",
    "Batch",
    "alert_transport_id",
    "balanced_partition_sizes",
    "batch_id_of",
    "build_batches",
    "group_alerts",
]

#: Hard ceiling from design section 5.1. Configuration may lower it, never raise it.
MAX_BATCH_SIZE: Final = 200


@dataclass(frozen=True, slots=True)
class AlertGroup:
    type: str
    """``alert_rule_url`` or ``application``."""
    value: str
    alerts: tuple[AlertRecord, ...]


@dataclass(frozen=True, slots=True)
class Batch:
    batch_id: str
    group_type: str
    group_value: str
    partition_index: int
    partition_count: int
    alerts: tuple[AlertRecord, ...]
    alert_ids: tuple[str, ...]
    """Parallel to ``alerts``."""


def alert_transport_id(alert: AlertRecord) -> str:
    """Transport identifier for one alert.

    Explicitly NOT a second business identity: it is the SHA-256 of an unambiguous encoding
    of (schema, application, key_field). The verdict cache stays keyed on
    (application, key_field, prompt_version, model_version) as decided.
    """
    return sha256_of(
        {
            "schema": alert.schema,
            "application": alert.application,
            "key_field": alert.key_field,
        }
    )


def batch_id_of(
    run_id: str,
    group_type: str,
    group_value: str,
    partition_index: int,
    alert_ids: Sequence[str],
    prompt_version: str,
    model_version: str,
) -> str:
    """Deterministic batch identifier covering membership, ordering and versions."""
    return sha256_of(
        {
            "run_id": run_id,
            "group_type": group_type,
            "group_value": group_value,
            "partition_index": partition_index,
            "alert_ids": list(alert_ids),
            "prompt_version": prompt_version,
            "model_version": model_version,
        }
    )


def _is_usable(value: object) -> bool:
    return isinstance(value, str) and value.strip() != ""


def group_alerts(alerts: Sequence[AlertRecord]) -> list[AlertGroup]:
    """Group alerts by rule URL, falling back to application.

    API alerts do not carry an alert-rule URL at all, which is why the fallback exists and
    why R4 does not penalise them for the absence.

    Groups are ordered by (type, value) and alerts within a group by key_field, with
    application breaking the tie so the order is total even when two applications share a
    key_field value.
    """
    grouped: dict[tuple[str, str], list[AlertRecord]] = {}
    for alert in alerts:
        if _is_usable(alert.alert_rule_url):
            key = ("alert_rule_url", str(alert.alert_rule_url))
        else:
            key = ("application", alert.application)
        grouped.setdefault(key, []).append(alert)

    return [
        AlertGroup(
            type=group_type,
            value=group_value,
            alerts=tuple(sorted(members, key=lambda a: (a.key_field, a.application))),
        )
        for (group_type, group_value), members in sorted(grouped.items())
    ]


def balanced_partition_sizes(n: int, maximum: int) -> list[int]:
    """Split ``n`` into ``ceil(n / maximum)`` partitions differing by at most one.

    401 alerts become 134, 134 and 133 - never 200, 200 and 1. A tiny tail request would
    lose exactly the same-group context that batching exists to provide.
    """
    if n <= 0:
        return []
    count = -(-n // maximum)  # ceil division
    base, remainder = divmod(n, count)
    return [base + 1 if i < remainder else base for i in range(count)]


def build_batches(
    alerts: Sequence[AlertRecord],
    run_id: str,
    prompt_version: str,
    model_version: str,
    max_batch_size: int | None = None,
) -> list[Batch]:
    """Build the batches for one run."""
    cap = min(max_batch_size or MAX_BATCH_SIZE, MAX_BATCH_SIZE)
    batches: list[Batch] = []

    for group in group_alerts(alerts):
        sizes = balanced_partition_sizes(len(group.alerts), cap)
        offset = 0
        for partition_index, size in enumerate(sizes):
            members = group.alerts[offset : offset + size]
            offset += size
            alert_ids = tuple(alert_transport_id(alert) for alert in members)
            batches.append(
                Batch(
                    batch_id=batch_id_of(
                        run_id,
                        group.type,
                        group.value,
                        partition_index,
                        alert_ids,
                        prompt_version,
                        model_version,
                    ),
                    group_type=group.type,
                    group_value=group.value,
                    partition_index=partition_index,
                    partition_count=len(sizes),
                    alerts=members,
                    alert_ids=alert_ids,
                )
            )

    return batches
